### PR TITLE
Ume datamodule, iterable datasets, concat and multiplexed datasets, throughput+tokens per second callback

### PR DESCRIPTION
**Pull request by karinazad**

### CODE REVIEW

File: src/lobster/hydra_config/experiment/train_ume.yaml
Comment (karinazad) [inline src/lobster/hydra_config/experiment/train_ume.yaml:1]: temporary, will remove before merging

File: src/lobster/assets/smiles_tokenizer/tokenizer_config.json [hunk 1]
```diff
@@ -1,31 +1,31 @@
 {
   "added_tokens_decoder": {
     "0": {
-      "content": "<pad>",
+      "content": "<cls>",
```
Comment (ncfrey) [inline src/lobster/assets/smiles_tokenizer/tokenizer_config.json:4]: this will actually be very nice and convenient to have special tokens have shared IDs across tokenizers

File: src/lobster/callbacks/_tokens_per_second.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            )
+
+
+def default_batch_size_fn(batch: Any) -> int:
+    """Default batch size function that returns the batch size."""
+    x = batch["input_ids"].squeeze(1)
+    return x.shape[0]
+
+
+def default_batch_length_fn(batch: Any) -> int:
```
Comment (ncfrey) [inline src/lobster/callbacks/_tokens_per_second.py:105]: do we need to account for padding/unpadding for flexbert?

Comment (ncfrey) [reply src/lobster/callbacks/_tokens_per_second.py:105]: probably ok just for profiling performance

File: src/lobster/data/_ume_datamodule.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        self._seed = seed
+        self._batch_size = batch_size
+        self._num_workers = num_workers
+        self._pin_memory = pin_memory
+        self._shuffle_buffer_size = shuffle_buffer_size
+
+        # Initialize tokenizer transforms for each modality
+        self._tokenizer_transforms = {
+            Modality.SMILES: TokenizerTransform(
+                SmilesTokenizerFast(), padding="max_length", truncation=True, max_length=self._tokenizer_max_length
```
Comment (ncfrey) [inline src/lobster/data/_ume_datamodule.py:60]: should we put the appropriate Tokenizer in `SUPPORTED_DATASETS_INFO` dict as well so it's all easily findable?

File: src/lobster/data/_ume_datamodule.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+from lobster.constants import Modality
+from lobster.datasets import AMPLIFYIterableDataset, CalmIterableDataset, M320MIterableDataset
+from lobster.tokenization import AminoAcidTokenizerFast, NucleotideTokenizerFast, SmilesTokenizerFast
+from lobster.transforms import TokenizerTransform
+
+# Supported datasets with modalities and supported splits
+SUPPORTED_DATASETS_INFO = {
+    "M320M": {"modality": Modality.SMILES, "supported_splits": ["train", "validation"]},
+    "Calm": {"modality": Modality.NUCLEOTIDE, "supported_splits": ["train"]},
```
Comment (ncfrey) [inline src/lobster/data/_ume_datamodule.py:16]: i'll update this, need train, val, test for all lol

File: src/lobster/data/_ume_datamodule.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        modality = dataset_info["modality"]
+        transform = self._tokenizer_transforms[modality]
+
+        # Check if the requested split is supported
+        if split not in dataset_info["supported_splits"]:
+            raise ValueError(f"Split '{split}' is not supported for dataset '{name}'")
+
+        match name:
+            case "M320M":
+                dataset = M320MIterableDataset(
```
Comment (ncfrey) [inline src/lobster/data/_ume_datamodule.py:86]: same as above, could specify the correct dataset class as well

File: src/lobster/datasets/_amplify_dataset.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
+from pathlib import Path
+from typing import Callable, ClassVar, Literal
+
+from lobster.datasets._huggingface_iterable_dataset import HuggingFaceIterableDataset
+from lobster.transforms import Transform
+
+
+class AMPLIFYIterableDataset(HuggingFaceIterableDataset):
+    """
+    Iterable Dataset for the AMPLIFY protein language model dataset (UR100P).
```
Comment (ncfrey) [inline src/lobster/datasets/_amplify_dataset.py:10]: this is a great docstring; should use this as a template for everything

File: src/lobster/hydra_config/experiment/train_ume.yaml
Comment (ncfrey) [reply src/lobster/hydra_config/experiment/train_ume.yaml:1]: probably good to have an example config here for people to use as a template

File: src/lobster/data/_ume_datamodule.py (hunk 3 above)
Comment (karinazad) [reply src/lobster/data/_ume_datamodule.py]: Added the dataset class to `SUPPORTED_DATASETS_INFO` and made it more structured. Tokenizers are a bit tricky there cause we probably want to reuse the instantiated tokenizers across modalities? For now, kept it separate

File: src/lobster/callbacks/_tokens_per_second.py (hunk 2 above)
Comment (karinazad) [reply src/lobster/callbacks/_tokens_per_second.py]: good idea! we can adjust that for flex bert specifically when unpad_embeddings=True

File: src/lobster/data/_ume_datamodule.py (hunk 4 above)
Comment (karinazad) [reply src/lobster/data/_ume_datamodule.py]: right now, I'm using the test sets for validation as well - is that ok? I was imagining that we'd test on completely different datasets?

File: src/lobster/data/_ume_datamodule.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+            dataset_info = next(info for info in SUPPORTED_DATASETS_INFO if info.name == dataset_name)
+
+            if Split.TRAIN in dataset_info.supported_splits:
+                train_dataset = self._get_dataset(dataset_info, split=Split.TRAIN)
+
+                self._train_datasets.append(train_dataset)
+                self._train_sizes.append(dataset_info.train_size)
+
+            # Uses test sets for validation
+            # This assumes that we're going to test on completely different datasets
```
Comment (ncfrey) [inline src/lobster/data/_ume_datamodule.py:158]: add TODO to standardize train, val, test in dataset creation pipeline

File: src/lobster/data/_ume_haha_datamodule.py [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+    HAHA = True
+
+except (ImportError, ModuleNotFoundError):
+    HAHA = False
+
+
+# Updated dictionary to include supported splits for each dataset
+SUPPORTED_DATASETS_INFO = {
+    "M320M": {"modality": Modality.SMILES, "supported_splits": ["train", "validation", "test"]},
+    "Calm": {"modality": Modality.NUCLEOTIDE, "supported_splits": ["train"]},
```
Comment (ncfrey) [inline src/lobster/data/_ume_haha_datamodule.py:26]: will fix in future MR

File: src/lobster/datasets/_multiplexed_sampling_dataset.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+class MultiplexedSamplingDataset(IterableDataset):
+    """Dataset that samples from multiple datasets according to specified weights."""
+
+    def __init__(
+        self,
+        datasets: Sequence[Dataset | IterableDataset],
+        weights: Sequence[float | int] = None,
+        seed: int | None = None,
+    ):
```
Comment (ncfrey) [inline src/lobster/datasets/_multiplexed_sampling_dataset.py:17]: add docstring

File: src/lobster/datasets/_shuffled_iterable_dataset.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+from torch.utils.data import IterableDataset, get_worker_info
+
+
+class ShuffledIterableDataset(IterableDataset):
+    def __init__(
+        self,
+        dataset: IterableDataset,
+        buffer_size: int = 10000,
+        seed: int | None = None,
+    ):
```
Comment (ncfrey) [inline src/lobster/datasets/_shuffled_iterable_dataset.py:12]: add docstring

File: src/lobster/datasets/_utils.py [hunk 11]
```diff
@@ -0,0 +1,32 @@
+import os
```
Comment (ncfrey) [inline src/lobster/datasets/_utils.py:1]: maybe rename to `_distributed_environment_utils.py`

File: src/lobster/hydra_config/model/modern_bert.yaml [hunk 12]
```diff
@@ -7,10 +7,6 @@ num_warmup_steps: 10_000
 max_length: 512
 tokenizer_dir: pmlm_tokenizer
 embedding_layer: linear_pos
-hidden_size: 252
```
Comment (ncfrey) [inline src/lobster/hydra_config/model/modern_bert.yaml:10]: do we need to drop these?

**Review by ncfrey — APPROVED** [review]

File: src/lobster/hydra_config/model/modern_bert.yaml (hunk 12 above)
Comment (karinazad) [reply src/lobster/hydra_config/model/modern_bert.yaml:10]: these were dropped in the config because when we provide the model config name (mini, medium), it would receive two hidden_size args
